core/src/test/java/org/apache/iceberg/rest/RESTCatalogServlet.java

```diff
@@ -0,0 +1,231 @@
+              context.method(),
+              context.path(),
+              context.queryParams(),
+              context.body(),
+              context.route().getResponseClass(),
+              context.headers(),
+              handle(response));
+
+      if (responseBody != null) {
+        RESTObjectMapper.mapper().writeValue(response.getWriter(), responseBody);
+      }
+    } catch (RESTException e) {
+      LOG.error("Error processing REST request", e);
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    } catch (Exception e) {
+      LOG.error("Unexpected exception when processing REST request", e);
+      response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    }
+  }
+
+  protected Consumer<ErrorResponse> handle(HttpServletResponse response) {
+    return (errorResponse) -> {
+      response.setStatus(errorResponse.code());
+      try {
+        RESTObjectMapper.mapper().writeValue(response.getWriter(), errorResponse);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    };
+  }
+
+  public static class ServletRequestContext {
+    private HTTPMethod method;
+    private Route route;
+    private String path;
+    private Map<String, String> headers;
+    private Map<String, String> queryParams;
+    private Object body;
+
+    private ErrorResponse errorResponse;
+
+    private ServletRequestContext(ErrorResponse errorResponse) {
+      this.errorResponse = errorResponse;
+    }
+
+    private ServletRequestContext(
+        HTTPMethod method,
+        Route route,
+        String path,
+        Map<String, String> headers,
+        Map<String, String> queryParams,
+        Object body) {
+      this.method = method;
+      this.route = route;
+      this.path = path;
+      this.headers = headers;
+      this.queryParams = queryParams;
+      this.body = body;
+    }
+
+    static ServletRequestContext from(HttpServletRequest request) throws IOException {
+      HTTPMethod method = HTTPMethod.valueOf(request.getMethod());
+      String path = request.getRequestURI().substring(1);
+      Pair<Route, Map<String, String>> routeContext = Route.from(method, path);
+
+      if (routeContext == null) {
+        return new ServletRequestContext(
+            ErrorResponse.builder()
+                .responseCode(400)
+                .withType("BadRequestException")
+                .withMessage(format("No route for request: %s %s", method, path))
+                .build());
+      }
+
+      Route route = routeContext.first();
+      Object requestBody = null;
+      if (route.getRequestClass() != null) {
+        requestBody =
+            RESTObjectMapper.mapper().readValue(request.getReader(), route.getRequestClass());
+      } else if (route == Route.TOKENS) {
+        try (Reader reader = new InputStreamReader(request.getInputStream())) {
+          Splitter.MapSplitter formSplitter = Splitter.on("&").withKeyValueSeparator("=");
```

Comment:
This can be a constant. Does it also make sense to put this into `RESTUtil` rather than embedding it here?

Reply:
Added decode (equivalent to existing encode) to `RESTUtil` along with corresponding test.